### ISSUE TITLE
Incorrect display of commanders on charts, some logic issues, and some values that shouldn't be there

### ISSUE THREAD

**Issue by tlongers** [security-force-monitor/sfm-cms#352]
Oddly, the dead dictator `Sani Abacha` was pulled in a parent units chart drawn at 8 June 2016 for `144 Battalion` ([on wwic](https://whowasincommand.com/en/organization/detail/1008/)). I've investigated it a bit, and whilst seems that the structure of the chart is correct,  commanders are being incorrectly assigned sometimes - there are cases where the dates don't match, and persons with non-command roles are appearing.  I've also spotted some values that shouldn't be there, for example end dates to `organization:organization` that are not in the import dataset, signalling an importer issue that needs to be looked at. Sorry for the length of this bug report

Here's the chart that set me off...

<img width="188" alt="whowasincommand" src="https://user-images.githubusercontent.com/2115933/32668125-a5173322-c634-11e7-8871-7110078efcd3.png">

The chart is drawn as at `8 June 2016`. The first thing I can't figure out is why the chart is drawn at this date. 

But working upwards from the bottom:

 * The immediate parent of `144 Brigade` on `8 June 2016` is correctly stated as `14 Brigade`, a relationship that runs from `24 July 2012` to `16 October 2016`. However, the commander of `14 Brigade` at this date `Lawrence Fejoku` and not `Sylvester Audu` as noted in the chart. What's unusual about the posting of `Fejoku` as commander of `144 Brigade` is that he only has a sole `YYYY` value of `2016` in `person:membership:date_first_cited`, which might be a source of the error.
 * The parent of `14 Brigade` at the chart date of `8 June 2016` is correctly stated as `82 Division` (running from `8 February 2011`  to `11 September 2017`). The commander is correctly stated as `Ibrahim Attahiru` (in command from  `30 April 2016` with no end date for that posting)
 * The parent of `82 Division` at `8 June 2016` is correctly stated as `Nigerian Army Headquarters` (running from `1975` to `11 September 2017`). However, the commander of `Nigerian Army Headquarters` on `8 June 2016` is `Tukur Yusufu Buratai` (running `4 August 2015` to `5 August 2017`) and not `Enobong Okon Udoh`, for whom there is neither a start nor end date for posting in `Nigerian Army Headquarters`.
 * The parent of `Nigeria Army Headquarters` on `8 June 2016` is correctly stated as `Nigeria Army` (running from `6 July 1994` with no end date). However, the commander of `Nigeria Army` at `8 June 2016` is `Tukur Yusufu Buratai` (`4 August 2015` to `5 August 2017`) and not `Kenneth Tobia Minimah` (`16 January 2014` to `3 August 2015`) as stated.
 * The parent of `Nigeria Army` on `8 June 2016` is  stated as  `Defence Headquarters`. This is both correct and incorrect. The parent/child relationship runs `6 July 1994`  to `2 June 2015`. Curiously, `Defence Headquarters` does not actually have a parent with a specified date range after `2 June 2015`. This is apparent in the way that the parent units chart is drawn for `Nigeria Army` ([link](https://whowasincommand.com/en/organization/detail/1129/#parent-units)):

<img width="756" alt="whowasincommand" src="https://user-images.githubusercontent.com/2115933/32667776-97ee04a6-c633-11e7-847a-70bba808d77f.png">

* Staying with this and moving upwards, the commander of `Defence Headquarters` is incorrectly shown as `John Zaruwa`. This is interesting too: `John Zarawu` has a start date  of `25 October 2014` but no end date, but is not in the `command` role - no role is specified in the data. However, there are two commanders in post after the earliest date that `John Zarawu` is affiliated with `Defence Headquarters`. `Alex Sabunduh Badeh` overlaps (`16 January 2014` to `12 July 2015`), `Abayomi Gabriel Olonisakin` starts and ends a posting (`13 July 2015` to `18 January 2016`) but neither has command on `8 June 2016`, so the chart should not display any command underformation (or `Unknown commander`).

* `Armed Forces of the Federal Republic of Nigeria` is correctly stated as the parent of of `Defence Headquarters` (running from `6 July 1994` with no end date). However, the commander is not `Ola Sa'ad Ibrahim` as stated (he runs in post `4 October 2012` to `16 January 2014`) - we don't have a data point for the commander of this unit on 8 June 2016 so the commander should be noted as `unknown` on the chart.

* The parent of the `Armed Forces of the Federal Republic of Nigeria` is correctly stated as `Armed Forces Council`, although this a relationship with neither start nor end date. 
However, the chart  posts  Nigeria's dictator `Sani Abacha` as commander  `Armed Forces Council` on `8 June 2016`, which is incorrect: his stint was `6 July 1994` to `8 June 1998` (the date of his death) The correct person in the chain should be `Muhammadu Buhari` (who served in that role from `29 May 2015` to `1 November 2016`).

* At the chart date of `8 June 2016` the parent of the `Armed Forces Council` is correctly stated as `President of the Federal Republic of Nigeria`, though the dates from `6 July 1994` to `2 June 2015`. BUT, in this case the import data has no `end date` for the parent relationship, so I'm not sure where the value of `2 June 2015` comes from. The person in role of `President of the Federal Republic of Nigeria` on the chart date of `6 June 2016` is `Muhammadu Buhari` and not `Abdulsalami Abubakar` as stated in the chart (he runs from `8 June 1998` to `28 May 1999`).

**Comment by jeancochrane** [comment]
Thanks for this detailed bug report @tlongers! I appreciate the thoroughness of your debugging.

Some thoughts:

1. The chart is drawn at `8 June 2016` because that's the end date we have on file for the composition between `144 Battalion` and its direct parent, `14 Brigade`. If that date doesn't match the composition end date in the sheet, we may have an importer bug on our hands :(
2. Recall that the way we calculate overlaps for this chart [isn't very sensitive to missing dates](https://github.com/security-force-monitor/sfm-cms/issues/223#issuecomment-337056269). I suspect this is behind at least some of the incorrect org-org command relationships here.
3. Taking a closer look at the query that generates the hierarchies for these charts, it looks like we're not identifying commanders based on the date that gets inputted, so it's almost certainly just picking the first commander it finds.

So basically, my diagnosis boils down to:

1. Importer error occasionally recording the wrong dates for compositions
2. Hierarchy query doesn't handle missing dates well
3. Hierarchy query assigns commanders randomly without considering posting dates

Do you think these three problems account for the issues you're seeing? Did I miss anything?

**Comment by tlongers** [comment]
Thanks Jean. The date `8 June 2016` is correct and taken from the correct location in the import sheet, which makes me think we don't have problem `1`.

I'll work through the logic of `Problem 2` a bit next week and see what cases we have and how we want the chart to respond in different cases.

I think `3. Hierarchy query assigns commanders randomly without considering posting dates` best describes the bulk of the issues in this report, so that should be a priority fix in my view.

**Comment by jeancochrane** [comment]
Great @tlongers, I'll focus on problem 3 for now then.

**Comment by tlongers** [comment]
@tonysecurityforcemonitor have you caught any chart weirdness in recent data updates that may indicate that this issue is still a live one? I think we've got a lot better on data content and import validation, which could reduce the incidence of this kind of error. I also wonder whether the approach that we have developed for graph querying in `sfm-build` may go some way to resolving the underlying idea of this issue, which is how to manage ambiguity in start/end dates. 

@hancush could you give me an idea of when the last time the chart code was touched, and whether you think this issue has been addressed?

**Comment by tlongers** [comment]
Quick update: The record for [96 Batallón de Infantería](https://whowasincommand.com/en/organization/view/10eadfc0-fe74-4e5e-bc79-05036e0c55d8/) of the Mexican Army be a good record to examine chart issues, as it has both very clear chains, and very broken, messy ones.
